stellargraph/layer/deep_graph_infomax.py

```diff
@@ -0,0 +1,162 @@
+
+__all__ = ["DeepGraphInfomax", "DGIDiscriminator"]
+
+
+class DGIDiscriminator(Layer):
+    """
+    This Layer computes the Discriminator function for Deep Graph Infomax (https://arxiv.org/pdf/1809.10341.pdf).
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def build(self, input_shapes):
+
+        first_size = input_shapes[0][-1]
+        second_size = input_shapes[1][-1]
+
+        self.kernel = self.add_weight(
+            shape=(first_size, second_size),
+            initializer="glorot_uniform",
+            name="kernel",
+            regularizer=None,
+            constraint=None,
+        )
+        self.built = True
+
+    def call(self, inputs):
+        """
+        Applies the layer to the inputs.
+
+        Args:
+            inputs: a list or tuple of tensors with shapes [(batch, N, F1), (batch, F2)] containing the node features and a
+                summary feature vector.
+        Returns:
+            a Tensor with shape (1, N)
+        """
+
+        features, summary = inputs
+
+        score = tf.linalg.matvec(features, tf.linalg.matvec(self.kernel, summary),)
+
+        return score
+
+
+@experimental(reason="lack of unit tests", issues=[1003])
+class DeepGraphInfomax:
+    """
+    A class to wrap stellargraph models for Deep Graph Infomax unsupervised training
+    (https://arxiv.org/pdf/1809.10341.pdf).
+
+    Args:
+        base_model: the base stellargraph model class
+    """
+
+    def __init__(self, base_model):
+
+        if not isinstance(base_model, (GCN, GAT, APPNP, PPNP)):
+            raise TypeError(
+                f"base_model: expected GCN, GAT, APPNP or PPNP found {type(base_model).__name__}"
+            )
+
+        if base_model.multiplicity != 1:
+            warnings.warn(
+                f"multiplicity: expected the base_model to have a multiplicity of 1, found"
+                f" ({self.base_model.multiplicity}). A multiplicity of 1 will be used to construct the base model."
+            )
+
+        self.base_model = base_model
+
+        self._node_feats = None
+        self._unique_id = hex(id(self))
+        # specific to full batch models
+        self._corruptible_inputs_idxs = [0]
+
+    def build(self):
+        """
+        A function to create the the keras inputs and outputs for a Deep Graph Infomax model for unsupervised training.
+
+        Note that the tf.nn.sigmoid_cross_entropy_with_logits loss must be used with this model.
+
+        Example::
+
+            dg_infomax = DeepGraphInfoMax(...)
+            x_in, x_out = dg_infomax.build()
+            model = Model(inputs=x_in, outputs=x_out)
+            model.compile(loss=tf.nn.sigmoid_cross_entropy_with_logits, ...)
+
+        Returns:
+            input and output layers for use with a keras model
+        """
+
+        x_inp, node_feats = self.base_model.build(multiplicity=1)
+        node_feats = Lambda(lambda x: x, name=self._unique_id)(node_feats)
```

Comment:
A brief comment would make it obvious why we're doing this mildly peculiar thing. E.g.

```suggestion
        # identity layer so we can attach a name to the tensor
        node_feats = Lambda(lambda x: x, name=self._unique_id)(node_feats)
```